tsdb/exemplar.go

```diff
@@ -0,0 +1,135 @@
+	lock      sync.RWMutex
+	index     map[string]int
+	exemplars []circularBufferEntry
+	nextIndex int
+	len       int
+}
+
+type circularBufferEntry struct {
+	exemplar     exemplar.Exemplar
+	seriesLabels labels.Labels // need to store labels so we can double check when querying
+	prev         int           // index of previous exemplar in circular for the same series
+}
+
+// If we assume the average case 95 bytes per exemplar we can fit 5651272 exemplars in
+// 1GB of extra memory, accounting for the fact that this is heap allocated space.
+func NewCircularExemplarStorage(len int) *CircularExemplarStorage {
+	return &CircularExemplarStorage{
+		exemplars: make([]circularBufferEntry, len),
+		index:     make(map[string]int),
+		len:       len,
+	}
+}
+
+func (ce *CircularExemplarStorage) Appender() storage.ExemplarAppender {
+	return ce
+}
+
+// TODO: separate wrapper struct for queries?
+func (ce *CircularExemplarStorage) Querier(ctx context.Context) (storage.ExemplarQuerier, error) {
+	return ce, nil
+}
+
+// Select returns exemplars for a given set of series labels hash.
+func (ce *CircularExemplarStorage) Select(hints *storage.SelectHints, l labels.Labels) ([]exemplar.Exemplar, error) {
+	var (
+		ret []exemplar.Exemplar
+		e   exemplar.Exemplar
+		idx int
+		ok  bool
+		buf []byte
+	)
+
+	ce.lock.RLock()
+	defer ce.lock.RUnlock()
+
+	if idx, ok = ce.index[l.String()]; !ok {
+		return nil, nil
+	}
+	lastTs := ce.exemplars[idx].exemplar.Ts
+
+	for {
+		if idx == -1 || string(ce.exemplars[idx].seriesLabels.Bytes(buf)) != string(l.Bytes(buf)) {
+			break
+		}
+
+		e = ce.exemplars[idx].exemplar
+		// This line is needed to avoid an infinite loop, consider redesign of buffer entry struct.
+		if e.Ts > lastTs {
+			break
+		}
+
+		lastTs = ce.exemplars[idx].exemplar.Ts
+		// Prepend since this exemplar came before the last one we appeneded chronologically.
+		if e.Ts >= hints.Start && e.Ts <= hints.End {
+			ret = append([]exemplar.Exemplar{e}, ret...)
+		}
+		idx = ce.exemplars[idx].prev
+	}
+	return ret, nil
+}
+
+func (ce *CircularExemplarStorage) AddExemplar(l labels.Labels, t int64, e exemplar.Exemplar) error {
+	ce.lock.RLock()
+	idx, ok := ce.index[l.String()]
+	ce.lock.RUnlock()
+
+	ce.lock.Lock()
+	defer ce.lock.Unlock()
+
+	if ok {
+		// Check for duplicate vs last stored exemplar for this series.
+		if ce.exemplars[idx].exemplar.Equals(e) {
+			return storage.ErrDuplicateExemplar
+		}
+		if e.Ts <= ce.exemplars[idx].exemplar.Ts {
+			return storage.ErrOutOfOrderExemplar
+		}
+		ce.exemplars[ce.nextIndex] = circularBufferEntry{exemplar: e, seriesLabels: l, prev: idx}
+		ce.index[l.String()] = ce.nextIndex
+		ce.nextIndex++
+		if ce.nextIndex >= cap(ce.exemplars) {
+			ce.nextIndex = 0
+		}
+		return nil
+	}
+	ce.exemplars[ce.nextIndex] = circularBufferEntry{exemplar: e, seriesLabels: l, prev: -1}
+	ce.index[l.String()] = ce.nextIndex
+	ce.nextIndex++
+	if ce.nextIndex >= cap(ce.exemplars) {
+		ce.nextIndex = 0
+	}
+	return nil
```

Comment:
We could keep this inside a `if !ok { }` and the other one outside as this is a smaller piece of code :)